File: chaine.rb
class Chaine

    attr_accessor :data, :prevChaine, :nextChaine, :firstChaine
    
    def initialize(data)
        @data = data
        @prevChaine = nil
        @nextChaine = nil
        @firstChaine = nil
    end

    def add(chaine)
        # tant que self.nextChaine != null
            # self.add(chaine)
        # sinon self.nextChaine = chaine
        if (@nextChaine != nil)
            @nextChaine.add(chaine)
        else
            @nextChaine = chaine
        end
    end

    def removeByData(data)
    end

    def removeByIndex(index)
    end

    def print
        puts data
        unless @nextChaine.nil?
            @nextChaine.print()
            puts @firstChaine.data
        else
            puts "END"
        end
    end

    def printAtFirst

    end

end

a = Chaine.new("la chaine A")

b = Chaine.new("la chaine B")

c = Chaine.new("la chaine C")

a.add(b)
a.add(c)

b.print()

File: chaine.js
class Chaine {

    constructor(data) {
        this.data = data
        this.prevChaine = null
        this.nextChaine = null
        this.firstChaine = null
    }

    add(chaine) {
        // si this.firstchaine && chaine.firstchaine == null
        //      this.firstchaine = this
        // dans tout les cas
        //     chaine.firstchaine = this.firstchaine
        if (this.firstChaine == null && chaine.firstChaine == null) {
            this.firstChaine = this
        } 
        chaine.firstChaine = this.firstChaine
        
        
        // tant que self.nextChaine != null
        // self.add(chaine)
        // sinon self.nextChaine = chaine
        if (this.nextChaine != null) {
            this.nextChaine.add(chaine)
        } else {
            this.nextChaine = chaine
        }

    }

    removeByData(data) {
    }

    removeByIndex(index) {
    }

    // at first
    print() {

        console.log(this)
       if (this.nextChaine)
        this.nextChaine.print()

    }



}

let a = new Chaine("la chaine A")
let b = new Chaine("la chaine B")
let c = new Chaine("la chaine C")

a.add(b)
a.add(c)

b.print()
